support-workers/src/typescript/lambdas/createPaymentMethodLambda.ts

```diff
@@ -0,0 +1,194 @@
+	paymentFields: PaymentFields,
+	user: User,
+): Promise<PaymentMethod> {
+	switch (paymentFields.paymentType) {
+		case 'Stripe':
+			return createStripePaymentMethod(user.isTestUser, paymentFields);
+		case 'PayPal':
+			return createPayPalPaymentMethod(user.isTestUser, paymentFields);
+		case 'DirectDebit':
+			return createDirectDebitPaymentMethod(paymentFields, user);
+		case 'Existing':
+			return Promise.reject(
+				new Error(
+					'Existing payment methods should never make their way to this lambda',
+				),
+			);
+	}
+	return Promise.reject(new Error('Unknown payment method type'));
+}
+
+export function createSalesforceContactState(
+	inputState: CreatePaymentMethodState,
+	paymentMethod: PaymentMethod,
+): WrappedState<CreateSalesforceContactState> {
+	const outputState: CreateSalesforceContactState = {
+		...inputState,
+		paymentMethod,
+	};
+
+	return {
+		state: outputState,
+		error: null,
+		requestInfo: {
+			testUser: false,
+			failed: false,
+			messages: [],
+			accountExists: false,
+		},
+	};
+}
+
+const zuoraCardTypeFromStripe = (stripeCardType: string): string => {
+	switch (stripeCardType) {
+		case 'visa':
+			return 'Visa';
+		case 'mastercard':
+			return 'MasterCard';
+		case 'amex':
+			return 'AmericanExpress';
+		case 'discover':
+			return 'Discover';
+		case 'jcb':
+			return 'JCB';
+		case 'diners':
+			return 'Diners';
+		default:
+			throw new Error(`Unknown card type ${stripeCardType}`);
+	}
+};
+
+export async function createStripePaymentMethod(
+	isTestUser: boolean,
+	paymentFields: StripePaymentFields,
+): Promise<StripePaymentMethod> {
+	const stripeService = await stripeServiceHandler.getServiceForUser(
+		isTestUser,
+	);
+	const customer = await stripeService.createCustomer(
+		paymentFields.stripePublicKey,
+		paymentFields.paymentMethod,
+	);
+	const paymentMethod = await stripeService.getPaymentMethod(
+		paymentFields.stripePublicKey,
+		paymentFields.paymentMethod,
+	);
+	const card = getIfDefined(
+		paymentMethod.card,
+		`Couldn't retrieve card details from Stripe for customer ${customer.id}`,
+	);
+	return {
+		TokenId: paymentFields.paymentMethod,
+		SecondTokenId: customer.id,
+		CreditCardNumber: card.last4,
+		CreditCardCountry: card.country,
+		CreditCardExpirationMonth: card.exp_month,
+		CreditCardExpirationYear: card.exp_year,
+		CreditCardType: zuoraCardTypeFromStripe(card.brand),
+		PaymentGateway: stripeService.getPaymentGateway(
+			paymentFields.stripePublicKey,
+		),
+		Type: 'CreditCardReferenceTransaction',
+		StripePaymentType: paymentFields.stripePaymentType,
+	};
+}
+
+async function createPayPalPaymentMethod(
+	isTestUser: boolean,
+	payPal: PayPalPaymentFields,
+): Promise<PayPalPaymentMethod> {
+	const payPalService = await paypalServiceHandler.getServiceForUser(
+		isTestUser,
+	);
+	const email = await payPalService.retrieveEmail(payPal.baid);
+	return {
+		PaypalBaid: payPal.baid,
+		PaypalEmail: getIfDefined(email, 'Could not retrieve email from PayPal'),
```

Comment:
I feel like validation of the response should be down to `payPalService.retrieveEmail` rather than trying to do it later, I've found the getIfDefined calls in the business logic clutter things up a lot.
I guess it's because `Either` is not really a thing in typescript but would throwing an Error/exception be a idiomatic alternative?